urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/ops/OutwardReduce.java

```diff
@@ -0,0 +1,763 @@
+			//Although remove the edge may increase less area. Remove edge is more expensive. It compute new point which 
+			//might introduce lots of error. Of course, usually, even the cost is computed, it is larger than that of concave vertex.
+			//we remove this case to simplify the algorithm.
+			return new ReduceEdge(start,end, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY,null);
+		}
+		Segment seg1 = Segment.create(start.prev, start);
+		Segment seg2 = Segment.create(end, end.next);
+		GeomObj2 obj = edgeIntersect(seg1,seg2);
+		try{
+			if(obj instanceof Point){//intersection is point
+				Point p = (Point)obj;
+				if(Point.isRightTurn(start, p, end)){//intersection in polygon, can't remove
+					area = Double.POSITIVE_INFINITY;
+					cost = Double.POSITIVE_INFINITY;
+				//NOTE After removing co-linear points, the intersection point can only be left/right turn. 
+				// However, it might be on the same line because of rounding off, which usually happens when start end and p are almost the same.
+				// We can remove such points because the error should be ignorable. 	
+				}else{	
+				//}else if(Point.isLeftTurn(start, p, end)){//intersection out of polygon, can remove
+					//update polyCost
+					//NOTE remove negative or 0 area because of round off
+					//NOTE: use the increase length to avoid sharp corner.
+					
+					//double areaDist[] = areaDist(start,p,end);
+					//area = Math.max(Double.MIN_VALUE, areaDist[0]);
+					area = Math.max(Double.MIN_VALUE, areaDist(start,p,end)[0]);
+
+					double d = segLen(start,end);
+					double d1 = segLen(start,p); double d2 = segLen(p,end);
+					double dist = d1+d2-d; 
+					//double dist = areaDist[1];
+					dist += Math.max(start.dist(),end.dist());
+					if(dist>maxNewPointDist){
+						cost = Double.POSITIVE_INFINITY;
+					}else{
+						cost = Math.max(area, dist*dist*hullWeight/2);
+					}
+					//NOTE: assign the dist of this pt to convex hull (over approximated).
+					obj = new ReduceVertex((Point)obj,dist);
+				//}else{//intersection on edge, impossible
+				//	//This may caused by floating point error.
+				//	throw new RuntimeException("Impossible, intersection is segment");
+				}
+			}else if(obj instanceof Empty){//parallel, can't remove
+				area = Double.POSITIVE_INFINITY;
+				cost = Double.POSITIVE_INFINITY;
+			}else if(obj instanceof Line){//on the same line, remove with no penalty
+				//NOTE: The special case of colinear point which can also be reduced by point. 
+				//Here, we change the cost from zero to infinity to force to use reducePoint. 
+				//That's because if we reduce by edge, we should handle this case specially because it reduce two points and doesn't introduce a new point.
+				//In the constructor we remove the colinear case however it's possible caused during the reduction
+				area = Double.POSITIVE_INFINITY;
+				cost = Double.POSITIVE_INFINITY;
+			}else{
+				throw new GeomException("Impossible");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [GeomException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [503].  Consider adding a catch clause of this specific exception in the outer try statement.